### PR TITLE
PLGN-394-Adding in new actions for get_user and get_asset information from an rrn

### PR DESCRIPTION
**Pull request by rbowden-r7**
### Description

Describe the proposed changes:
  - to add in the following for getting user info and asset information from an rrn 
  - add in unit tests to test the new actions 
  - to ensure that all headers are correctly set before each network call, as if the headers get changed from the default they are never reset before a new call that uses the default headers 
  - to update the spec file to ensure all actions have examples

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Testing

testing the new functionality by adding new integration tests - https://github.com/rapid7/icon-plugin-integration-tests/pull/163

testing the new actions using local docker image - https://github.com/rapid7/icon-plugin-integration-tests/pull/163

#### Unit Tests

new unit tests have been added to test the new actions that have been added

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/rapid7_insightidr/help.md [hunk 1]
```diff
@@ -47,796 +47,919 @@ Example input:
...
-  ],
+  "domain_names": "rapid7.com",
+  "hashes": "A94A8FE5CCB19BA61C4C0873D391E987982FBBD3",
+  "ips": "10.0.0.1",
   "key": "c9404e11-b81a-429d-9400-05c531f229c3",
-  "urls": [
-    "https://example.com",
-    "https://test.com"
-  ]
+  "urls": "https://example.com"
```
Comment (llaszuk-r7) [inline plugins/rapid7_insightidr/help.md:73]: same

File: plugins/rapid7_insightidr/help.md [hunk 2]
```diff
@@ -47,796 +47,919 @@ Example input:
...
-  "hashes": [
-    "A94A8FE5CCB19BA61C4C0873D391E987982FBBD3",
-    "C3499C2729730A7F807EFB8676A92DCB6F8A3F8F"
-  ],
-  "ips": [
-    "10.0.0.1",
-    "10.0.0.2"
-  ],
+  "domain_names": "rapid7.com",
+  "hashes": "A94A8FE5CCB19BA61C4C0873D391E987982FBBD3",
```
Comment (llaszuk-r7) [inline plugins/rapid7_insightidr/help.md:70]: same

File: plugins/rapid7_insightidr/help.md [hunk 3]
```diff
@@ -47,796 +47,919 @@ Example input:
...
-  ],
-  "hashes": [
-    "A94A8FE5CCB19BA61C4C0873D391E987982FBBD3",
-    "C3499C2729730A7F807EFB8676A92DCB6F8A3F8F"
-  ],
-  "ips": [
-    "10.0.0.1",
-    "10.0.0.2"
-  ],
+  "domain_names": "rapid7.com",
```
Comment (llaszuk-r7) [inline plugins/rapid7_insightidr/help.md:69]: it should be a list of strings, please make sure that our insight-plugin didn't break old examples

File: plugins/rapid7_insightidr/help.md [hunk 4]
```diff
@@ -47,796 +47,919 @@ Example input:
...
+  "hashes": "A94A8FE5CCB19BA61C4C0873D391E987982FBBD3",
+  "ips": "10.0.0.1",
```
Comment (llaszuk-r7) [inline plugins/rapid7_insightidr/help.md:71]: same

File: plugins/rapid7_insightidr/Dockerfile [hunk 5]
```diff
@@ -1,24 +1,18 @@
-FROM rapid7/insightconnect-python-3-38-slim-plugin:4
-# Refer to the following documentation for available SDK parent images: https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide
+FROM rapid7/insightconnect-python-3-38-plugin:4
```
Comment (llaszuk-r7) [inline plugins/rapid7_insightidr/Dockerfile:1]: update to 5, please

**Review by llaszuk-r7 — CHANGES_REQUESTED** [review]
Please check the examples in help.md, I found a few inconsistencies when comparing old examples, if you haven't changed any types the examples should stay in the same format.

File: plugins/rapid7_insightidr/help.md (hunk 1 above)
Comment (rbowden-r7) [reply plugins/rapid7_insightidr/help.md:73]: Yeah good catch that was changed by the refresh tool, but has now been revert ed back to a list

File: plugins/rapid7_insightidr/help.md (hunk 2 above)
Comment (rbowden-r7) [reply plugins/rapid7_insightidr/help.md:70]: Yeah good catch that was changed by the refresh tool, but has now been revert ed back to a list

File: plugins/rapid7_insightidr/help.md (hunk 3 above)
Comment (rbowden-r7) [reply plugins/rapid7_insightidr/help.md:69]: Yeah good catch that was changed by the refresh tool, but has now been revert ed back to a list

File: plugins/rapid7_insightidr/help.md (hunk 4 above)
Comment (rbowden-r7) [reply plugins/rapid7_insightidr/help.md:71]: Yeah good catch that was changed by the refresh tool, but has now been reverted back to a list

File: plugins/rapid7_insightidr/Dockerfile (hunk 5 above)
Comment (rbowden-r7) [reply plugins/rapid7_insightidr/Dockerfile:1]: this has been updated to use the newer sdk version

File: plugins/rapid7_insightidr/komand_rapid7_insightidr/actions/get_user_information/action.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+            output=GetUserInformationOutput(),
+        )
+
+    def run(self, params={}):
+        user_rrn = params.get(Input.USER_RRN)
+        self.connection.session.headers["Accept-version"] = "strong-force-preview"
+        request = ResourceHelper(self.connection.session, self.logger)
+        self.logger.info(f"Getting the user information for {user_rrn}...")
+        response = request.make_request(Users.get_user_information(self.connection.url, user_rrn), "get")
+        return {Output.USER: response, Output.SUCCESS: True}
```
Comment (igorski-r7) [inline plugins/rapid7_insightidr/komand_rapid7_insightidr/actions/get_user_information/action.py:24]: I am wondering if thats necessary to return the `Output.SUCCESS` while in the InsightConnect there is the same variable that returns if the action passed or not 😄

**Review by igorski-r7 — COMMENTED** [review]
Please make sure that all actions runu fine 😄 The new tooling sometimes breaks the `schema.py` files so that the action fails on the JSON schema validation.

File: plugins/rapid7_insightidr/komand_rapid7_insightidr/actions/get_user_information/action.py (hunk 6 above)
Comment (rbowden-r7) [reply plugins/rapid7_insightidr/komand_rapid7_insightidr/actions/get_user_information/action.py:24]: I had done it that was as I was following the other actions in that repo - https://github.com/rapid7/insightconnect-plugins/blob/db4ecd2117b9791baadcff023d808251e9f07273/plugins/rapid7_insightidr/komand_rapid7_insightidr/actions/get_attachment_information/action.py#L25

example of response returned below - I can remove this if needed as the results will still be very similar 
image

File: plugins/rapid7_insightidr/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (rbowden-r7) [reply plugins/rapid7_insightidr/Dockerfile]: this has been updated

**Comment by rbowden-r7** [comment]
> Please make sure that all actions runu fine 😄 The new tooling sometimes breaks the `schema.py` files so that the action fails on the JSON schema validation.

I have manually ran the the few actions that are not covered by the integration tests and they seem to run as expected

**Review by llaszuk-r7 — APPROVED** [review]

**Review by igorski-r7 — APPROVED** [review]

**Pull request merged**